### PR TITLE
convert db lib to use slog, part 2

### PR DESCRIPTION
**Pull request by Tener**
Follow up to https://github.com/gravitational/teleport/pull/43382.

The main types affected are: `Engine`, `EngineConfig` and `SessionContext`, with cascading effects on the rest of the code.

The changes are _mostly_ mechanical, but a few small refactors are also in.

### CODE REVIEW

File: lib/srv/db/mongodb/engine.go [hunk 1]
```diff
@@ -206,7 +207,7 @@ func (e *Engine) processHandshakeResponse(ctx context.Context, respMessage proto
 	case *protocol.MessageOpMsg:
 		rawMessage = bson.Raw(resp.BodySection.Document)
 	default:
-		e.Log.Warn("Unabled to process MongoDB handshake response. Unexpected message type %T", respMessage)
+		e.Log.WarnContext(ctx, "Unable to process MongoDB handshake response. Unexpected message type.", "message_type", fmt.Sprintf("%T", respMessage))
```
Comment (greedy52) [inline lib/srv/db/mongodb/engine.go:210]: use `lib/utils/log.TypeAttr`?

File: lib/srv/db/elasticsearch/util.go [hunk 2]
```diff
@@ -37,50 +37,52 @@ func GetQueryFromRequestBody(e common.EngineConfig, contentType string, body []b
 		Knn   any `json:"knn" yaml:"knn"`
 	}
 
+	log := e.Log.With("content-type", contentType)
```
Comment (greedy52) [inline lib/srv/db/elasticsearch/util.go:40]: nit: inline attributes are preferred unless `log` will be reused.

File: lib/srv/db/postgres/engine.go [hunk 3]
```diff
@@ -202,7 +202,7 @@ func (e *Engine) handleStartup(client *pgproto3.Backend, sessionCtx *common.Sess
 	}
 	switch m := startupMessageI.(type) {
 	case *pgproto3.StartupMessage:
-		e.Log.Debugf("Received startup message: %#v.", m)
+		e.Log.DebugContext(e.Context, "Received startup message.", "message", fmt.Sprintf("%#v", m))
```
Comment (greedy52) [inline lib/srv/db/postgres/engine.go:205]:
```suggestion
		e.Log.DebugContext(e.Context, "Received startup message.", "message", m)
```
slog prints out the go fields by default.

File: lib/srv/db/elasticsearch/util.go [hunk 4]
```diff
@@ -37,50 +37,52 @@ func GetQueryFromRequestBody(e common.EngineConfig, contentType string, body []b
...
 		err := json.Unmarshal(body, &q)
 		if err != nil {
-			e.Log.WithError(err).Warnf("Error decoding request body as %q.", contentType)
+			e.Log.WarnContext(e.Context, "Error decoding request body.", "error", err)
 			return ""
 		}
 
 	default:
-		e.Log.Warnf("Unknown or missing 'Content-Type': %q, assuming 'application/json'.", contentType)
+		log.WarnContext(e.Context, "Unknown or missing 'Content-Type',, assuming 'application/json'.")
```
Comment (codingllama) [inline lib/srv/db/elasticsearch/util.go]:
```suggestion
		log.WarnContext(e.Context, "Unknown or missing 'Content-Type', assuming 'application/json'.")
```

File: lib/srv/db/server.go [hunk 5]
```diff
@@ -1171,11 +1171,8 @@ func (s *Server) authorize(ctx context.Context) (*common.Session, error) {
 		AuthContext:        authContext,
 		Checker:            authContext.Checker,
 		StartupParameters:  make(map[string]string),
-		Log: s.logrusLogger.WithFields(logrus.Fields{
-			"id": id,
-			"db": database.GetName(),
-		}),
-		LockTargets: authContext.LockTargets(),
+		Log:                s.log.With("id", id, "db", database.GetName()),
```
Comment (greedy52) [inline lib/srv/db/server.go:1174]: 🎉

**Review by greedy52 — APPROVED** [review]
awesome! thanks!

File: lib/srv/db/mongodb/engine.go (hunk 1 above)
Comment (codingllama) [inline lib/srv/db/mongodb/engine.go:210]:
```suggestion
		e.Log.WarnContext(ctx, "Unable to process MongoDB handshake response. Unexpected message type.", "message_type", log.TypeAttr(respMessage))
```

log is lib/utils/log

File: lib/srv/db/mongodb/autousers_admin.go [hunk 6]
```diff
@@ -156,7 +157,7 @@ func newShareableAdminClient(cfg shareableAdminClientConfig) (*shareableAdminCli
 }
 
 func (c *shareableAdminClient) waitAndCleanup() {
-	c.log.Debugf("Created new MongoDB connection as admin user %q on %q.", c.adminUser, c.databaseName)
+	c.log.Debug("Created new MongoDB admin connection.", "user", c.adminUser, "database", c.databaseName)
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/srv/db/mongodb/autousers_admin.go:160]: Ah, we will see. There wasn't easily accessible context here, but I can thread one through if needed.

File: lib/srv/db/mysql/autousers.go [hunk 7]
```diff
@@ -360,18 +360,18 @@ func checkRoles(conn *clientConn, roles []string) error {
...
+	return trace.Wrap(checkMySQLSupportedVersion(log, conn.GetServerVersion()))
 }
 
-func checkMySQLSupportedVersion(serverVersion string) error {
+func checkMySQLSupportedVersion(log *slog.Logger, serverVersion string) error {
 	ver, err := semver.NewVersion(serverVersion)
 	switch {
 	case err != nil:
-		logrus.Debugf("Invalid MySQL server version %q. Assuming role management is supported.", serverVersion)
+		log.Debug("Invalid MySQL server version. Assuming role management is supported.", "server_version", serverVersion)
```
Comment (codingllama) [inline lib/srv/db/mysql/autousers.go:374]: Not DebugContext? Won't that fail linting?

File: lib/srv/db/elasticsearch/util.go (hunk 2 above)
Comment (Tener) [reply lib/srv/db/elasticsearch/util.go:40]: It is reused, though?

File: lib/srv/db/mysql/autousers.go [hunk 8]
```diff
@@ -401,7 +401,7 @@ func checkMariaDBSupportedVersion(serverVersion string) error {
 	ver, err := semver.NewVersion(serverVersion)
 	switch {
 	case err != nil:
-		logrus.Debugf("Invalid MariaDB server version %q. Assuming role management is supported.", serverVersion)
+		log.Debug("Invalid MariaDB server version. Assuming role management is supported.", "server_version", serverVersion)
```
Comment (codingllama) [inline lib/srv/db/mysql/autousers.go:404]: Same here for DebugContext.

File: lib/srv/db/postgres/engine.go (hunk 3 above)
Comment (codingllama) [inline lib/srv/db/postgres/engine.go]: We don't have a LogValuer for `%#v` (that I'm aware of), although I suppose we could make one.

File: lib/srv/db/postgres/engine.go [hunk 9]
```diff
@@ -313,19 +313,18 @@ func (e *Engine) makeClientReady(client *pgproto3.Backend, hijackedConn *pgconn.
...
 
 	for {
 		message, err := client.Receive()
 		if err != nil {
-			log.WithError(err).Errorf("Failed to receive message from client.")
+			log.ErrorContext(e.Context, "Failed to receive message from client.", "error", err)
 			clientErrCh <- err
 			return
 		}
-		log.Tracef("Received client message: %#v.", message)
```
Comment (codingllama) [inline lib/srv/db/postgres/engine.go:328]: FYI: You can log traces using log.TraceLevel (also from lib/utils/log).

File: lib/srv/db/postgres/engine.go [hunk 10]
```diff
@@ -448,23 +447,22 @@ func (e *Engine) receiveFromServer(serverConn *pgconn.PgConn, serverErrCh chan<-
...
 					return
 				}
-				log.WithError(err).Error("Failed to receive message from server.")
+				log.ErrorContext(e.Context, "Failed to receive message from server.", "error", err)
 				return
 			}
 
 			count += 1
 			ctr.Inc()
-			log.Tracef("Received server message: %#v.", message)
```
Comment (codingllama) [inline lib/srv/db/postgres/engine.go:467]: Same here for trace logging, if you want.

File: lib/srv/db/postgres/users.go [hunk 11]
```diff
@@ -286,13 +287,13 @@ func (e *Engine) removePermissions(ctx context.Context, sessionCtx *common.Sessi
...
 	if err != nil {
-		e.Log.WithError(err).Errorf("Creating temporary stored procedure %q failed.", removePermissionsProcName)
+		logger.ErrorContext(e.Context, "Creating temporary stored procedure failed.", "procedure", removePermissionsProcName, "error", err)
 		return trace.Wrap(err)
 	}
 
 	_, err = conn.Exec(ctx, removePermissionsQuery, sessionCtx.DatabaseUser)
 	if err != nil {
-		e.Log.WithError(err).WithField("user", sessionCtx.DatabaseUser).Error("Removing permissions from user failed.")
+		logger.Error("Removing permissions from user failed.", "error", err)
```
Comment (codingllama) [inline lib/srv/db/postgres/users.go:296]: Not ErrorContext?

File: lib/srv/db/sqlserver/engine.go [hunk 12]
```diff
@@ -131,12 +131,12 @@ func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Sessio
 func (e *Engine) receiveFromClient(clientConn, serverConn io.ReadWriteCloser, clientErrCh chan<- error, sessionCtx *common.Session) {
 	defer func() {
 		if r := recover(); r != nil {
-			e.Log.Warnf("Recovered while handling DB connection %v", r)
+			e.Log.ErrorContext(e.Context, "Recovered while handling DB connection", "recover", r)
```
Comment (codingllama) [inline lib/srv/db/sqlserver/engine.go:134]: Warn turned into Error?

```suggestion
			e.Log.WarnContext(e.Context, "Recovered while handling DB connection", "recover", r)
```

**Review by codingllama — APPROVED** [review]
Thanks for doing this!

Concurrent review, I might have some duplicate comments.

File: lib/srv/db/mongodb/autousers_admin.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Tener) [reply lib/srv/db/mongodb/autousers_admin.go:160]: Yup, plenty of complaints about missing context. Most of them in places I simply missed, as it happens. I'll add it.

File: lib/srv/db/postgres/engine.go (hunk 3 above)
Comment (codingllama) [reply lib/srv/db/postgres/engine.go]: TIL.

File: lib/srv/db/sqlserver/engine.go (hunk 12 above)
Comment (Tener) [reply lib/srv/db/sqlserver/engine.go:134]: Yes, that one is intentional. I think panics deserve higher severity than warn.

**Comment by Tener** [comment]
Thanks for the quick reviews!

**Comment by Tener** [comment]
@GavinFrazar @gabrielcorado are you planning on merging these two to v16:
- https://github.com/gravitational/teleport/pull/43335
- https://github.com/gravitational/teleport/pull/43546

It would lead to a cleaner merge of this PR to v16. If not then I can do manual merge just fine, but I'd rather coordinate this with you.

**Pull request merged**